File: backend/notes/Accidental.ts
class Accidental {
  private readonly offset: number

  private constructor(offset: number) {
    this.offset = offset
  }

  get displayString() {
    if (this.offset === 0) {
      return ''
    }
    if (this.offset < 0) {
      return 'b'.repeat(-this.offset)
    }
    if (this.offset % 2 === 0) {
      return 'x'.repeat(this.offset / 2)
    }
    return 'x'.repeat((this.offset - 1) / 2) + '#'
  }

  get sharp() {
    return new Accidental(this.offset + 1)
  }
  
  get flat() {
    return new Accidental(this.offset - 1)
  }

  static NATURAL = new Accidental(0)
  static SHARP = new Accidental(1)
  static DOUBLE_SHARP = new Accidental(2)
  static FLAT = new Accidental(-1)
  static DOUBLE_FLAT = new Accidental(-2)

}

export default Accidental

File: backend/src/main/kotlin/com/skewwhiffy/notation/model/note/Accidental.kt
package com.skewwhiffy.notation.model.note

data class Accidental(private val offset: Int) {
  companion object {
    val natural: Accidental = Accidental(0)
    val flat: Accidental = Accidental(-1)
    val sharp: Accidental = Accidental(1)
  }

  val abc = when {
    offset == 0 -> ""
    offset < 0 -> "_".repeat(-offset)
    else -> "^".repeat(offset)
  }

  val displayString: String = when {
    offset < 0 -> "b".repeat(-offset)
    offset % 2 == 0 -> "x".repeat(offset / 2)
    else -> "x".repeat((offset - 1) / 2) + "#"
  }

  val flat: Accidental
    get() = Accidental(offset - 1)

  val sharp: Accidental
    get() = Accidental(offset + 1)
}